standard/expressions.md

```diff
@@ -2221,29 +2218,94 @@ The *primary_expression* of an *element_access* shall not be an *array_creation_
+
+#### §string-access String access
+
+For a string access the *argument_list* of the *element_access* shall contain a single unnamed value argument (§15.6.2.2) which shall be:
+
+- of type `int`, `Index` or `Range`; or
+- implicitly convertible to one or more of the above types.
+
+The run-time processing of a string access of the form `P[A]`, where `P` is a *primary_expression* of `string` type and `A` is a single expression, consists of the following steps:
 
 - `P` is evaluated. If this evaluation causes an exception, no further steps are executed.
-- The index expressions of the *argument_list* are evaluated in order, from left to right. Following evaluation of each index expression, an implicit conversion ([§10.2](conversions.md#102-implicit-conversions)) to one of the following types is performed: `int`, `uint`, `long`, `ulong`. The first type in this list for which an implicit conversion exists is chosen. For instance, if the index expression is of type `short` then an implicit conversion to `int` is performed, since implicit conversions from `short` to `int` and from `short` to `long` are possible. If evaluation of an index expression or the subsequent implicit conversion causes an exception, then no further index expressions are evaluated and no further steps are executed.
+- The index expression is evaluated, let the type of the resultant value be *T*;
+- This value is then converted to the first of the types: `int`, `Index` or `Range`; for which an implicit conversion ([§10.2](conversions.md#102-implicit-conversions)) from *T* exists.
+- If evaluation of an index expression or the subsequent implicit conversion causes an exception, then no further index expressions are evaluated and no further steps are executed.
 - The value of `P` is checked to be valid. If the value of `P` is `null`, a `System.NullReferenceException` is thrown and no further steps are executed.
-- The value of each expression in the *argument_list* is checked against the actual bounds of each dimension of the array instance referenced by `P`. If one or more values are out of range, a `System.IndexOutOfRangeException` is thrown and no further steps are executed.
-- The location of the array element given by the index expressions is computed, and this location becomes the result of the array access.
+- If the preceding steps have produced an index value of type `Range` then:
+  - The result of evaluating the string access is a value of `string` type.
+  - Let *L* be the length of the string referenced by `P`.
+  - `A` is checked to be valid with respect to *L* (§24.3), if it is not then a `System.ArgumentOutOfRangeException` is thrown and no further steps are executed.
+  - The starting offset, *S*, and number of items, *N*, for `A` with respect to *L* are determined as described for `GetOffsetAndLength` (§24.3).
+  - The result of the string access is a new string formed by copying the *N* characters of `P` starting from *S*, if *N* is zero the new string is empty.
```

Comment:
Similarly to the array case, when the length of the range is zero, a new string need not be formed. The current compiler defers to String.Substring which returns a cached empty string instance when then length turns out to be zero at runtime (https://github.com/dotnet/runtime/blob/v9.0.8/src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs#L2241-L2256).

Reply:
Though the Standard does require *literal* strings in the *same* assembly that are equal, and two empty strings are clearly equal, to be the same object it does not require this for non-literal strings. If this issue is to be addressed it needs to be done globally and not just in this clause. PR #1403 has been spun off and this thread will be marked as resolved.